### PR TITLE
US881996: Replace gosu with su-exec

### PR DESCRIPTION
**Pull request by raghavendran-gopalakrishnan**
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=889116

### CODE REVIEW

**Comment by raghavgopal** [comment]
Created a `test_user`,

```
83d39e5df02e:/ # cut -d: -f1 /etc/passwd
root
test_user
```
Test with `su-exec':
```
83d39e5df02e:/ # su-exec test_user date
Tue 27 Feb 2024 02:46:46 PM UTC
```

Test with symlink `gosu`:
```
83d39e5df02e:/ # gosu test_user date
Tue 27 Feb 2024 02:46:51 PM UTC
```

**Comment by michael-bryson** [comment]
@raghavendran-gopalakrishnan The release notes and the main README.md need updated.

File: src/main/docker/Dockerfile [hunk 1]
```diff
@@ -42,35 +42,36 @@ RUN zypper -n refresh && \
...
-# Download and verify gosu
-RUN gpg --batch --keyserver-options http-proxy=${env.HTTP_PROXY} --keyserver hkps://keys.openpgp.org \
-        --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4 && \
-    curl -o /usr/local/bin/gosu -SL "https://github.com/tianon/gosu/releases/download/1.17/gosu-amd64" && \
-    curl -o /usr/local/bin/gosu.asc -SL "https://github.com/tianon/gosu/releases/download/1.17/gosu-amd64.asc" && \
-    gpg --batch --verify /usr/local/bin/gosu.asc /usr/local/bin/gosu && \
-    chmod +x /usr/local/bin/gosu
+# Download and compile su-exec
+RUN zypper -n refresh && \
+    zypper -n update && \
```
Comment (dermot-hardy) [inline src/main/docker/Dockerfile]: Skip refresh and update here since we just did it in stage 1.

File: src/main/docker/Dockerfile [hunk 2]
```diff
@@ -42,35 +42,36 @@ RUN zypper -n refresh && \
...
+    zypper -n install gcc && \
+    curl -o /usr/local/bin/su-exec.c -SL "https://raw.githubusercontent.com/ncopa/su-exec/master/su-exec.c" && \
```
Comment (dermot-hardy) [inline src/main/docker/Dockerfile:51]: With a direct download from the internet like this we need to verify that what has been downloaded is what we expect, i.e. by checking it's digest.

Here is an example where we are using `sha512sum -c` to do this:
https://github.com/CAFapi/buildenv-jdk11-image/blob/v1.6.3/src/main/docker/Dockerfile#L44-L45

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: src/main/docker/Dockerfile [hunk 3]
```diff
@@ -42,35 +42,36 @@ RUN zypper -n refresh && \
...
-        --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4 && \
-    curl -o /usr/local/bin/gosu -SL "https://github.com/tianon/gosu/releases/download/1.17/gosu-amd64" && \
-    curl -o /usr/local/bin/gosu.asc -SL "https://github.com/tianon/gosu/releases/download/1.17/gosu-amd64.asc" && \
-    gpg --batch --verify /usr/local/bin/gosu.asc /usr/local/bin/gosu && \
-    chmod +x /usr/local/bin/gosu
+# Download and compile su-exec
+RUN zypper -n install gcc && \
+    curl -o /usr/local/bin/su-exec.c -SL "https://raw.githubusercontent.com/ncopa/su-exec/master/su-exec.c" && \
+    echo 44d84566f8685273335fc44a82348eb71a6759792c9f0947f7a8be45dbac4d5685e7e6b59cdaf19bd13932f08bb383d9bd4aa37ff56ffac733695e55b5d76391 \
+    /usr/local/bin/su-exec.c | sha512sum -c - && \
```
Comment (dermot-hardy) [inline src/main/docker/Dockerfile]: Indent this a bit more.  The way it is it looks like another command rather than part of the previous one.

**Review by dermot-hardy — APPROVED** [review]

File: src/main/docker/Dockerfile (hunk 3 above)
Comment (raghavgopal) [reply src/main/docker/Dockerfile]: Addressed.

**Comment by raghavgopal** [comment]
@michael-bryson , please review the README and release notes.

File: release-notes-2.10.0.md [hunk 4]
```diff
@@ -4,5 +4,8 @@
 ${version-number}
 
 #### New Features
+- **US881996**: Replace gosu with su-exec  
+  - `gosu` utility has been replaced with a much similar and simpler `su-exec`.
```
Comment (michael-bryson) [inline release-notes-2.10.0.md]: This doesn't really make sense, I would just write:
```suggestion
  - The `gosu` utility has been replaced with `su-exec`.
```

File: release-notes-2.10.0.md [hunk 5]
```diff
@@ -4,5 +4,8 @@
 ${version-number}
 
 #### New Features
+- **US881996**: Replace gosu with su-exec
```
Comment (michael-bryson) [inline release-notes-2.10.0.md]:
```suggestion
- US881996: Replaced `gosu` with `su-exec`  
```

**Review by michael-bryson — APPROVED** [review]
Couple of small suggestions, otherwise approved.

**Pull request merged**